File: flutter_app/lib/models/plan.dart
import 'package:json_annotation/json_annotation.dart';

part 'plan.g.dart';

double _toDouble(dynamic v) =>
    v is num ? v.toDouble() : double.parse(v.toString());

@JsonSerializable()
class Plan {
  final String id;
  @JsonKey(name: 'country_id')
  final int countryId;
  final String tier;
  @JsonKey(name: 'name_en')
  final String nameEn;
  @JsonKey(name: 'name_ar')
  final String nameAr;
  @JsonKey(name: 'price_local', fromJson: _toDouble)
  final double priceLocal;
  @JsonKey(name: 'daily_rate', fromJson: _toDouble)
  final double dailyRate;
  @JsonKey(name: 'max_visits')
  final int maxVisits;
  @JsonKey(name: 'validity_days')
  final int validityDays;
  @JsonKey(name: 'duration_months')
  final int durationMonths;
  @JsonKey(name: 'gym_tier_access')
  final String gymTierAccess;
  @JsonKey(name: 'features_en')
  final List<dynamic>? featuresEn;
  @JsonKey(name: 'features_ar')
  final List<dynamic>? featuresAr;
  @JsonKey(name: 'is_active')
  final bool isActive;
  @JsonKey(name: 'sort_order')
  final int sortOrder;

  const Plan({
    required this.id,
    required this.countryId,
    required this.tier,
    required this.nameEn,
    required this.nameAr,
    required this.priceLocal,
    required this.dailyRate,
    this.maxVisits = 30,
    this.validityDays = 30,
    this.durationMonths = 1,
    required this.gymTierAccess,
    this.featuresEn,
    this.featuresAr,
    this.isActive = true,
    this.sortOrder = 0,
  });

  String name(String lang) => lang == 'ar' ? nameAr : nameEn;
  List<String> features(String lang) =>
      (lang == 'ar' ? featuresAr : featuresEn)?.cast<String>() ?? [];

  factory Plan.fromJson(Map<String, dynamic> json) => _$PlanFromJson(json);
  Map<String, dynamic> toJson() => _$PlanToJson(this);
}

File: backend/app/models/plan.py
import uuid
from sqlalchemy import Column, String, Integer, Boolean, DECIMAL, DateTime, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from app.database import Base


class Plan(Base):
    __tablename__ = "plans"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    country_id = Column(Integer, nullable=False)
    tier = Column(String(20), nullable=False)
    name_en = Column(String(100), nullable=False)
    name_ar = Column(String(100), nullable=False)
    price_local = Column(DECIMAL(12, 3), nullable=False)
    daily_rate = Column(DECIMAL(12, 3), nullable=False)
    max_visits = Column(Integer, nullable=False, default=30)
    validity_days = Column(Integer, nullable=False, default=30)
    duration_months = Column(Integer, nullable=False, default=1)
    gym_tier_access = Column(String(20), nullable=False)
    features_en = Column(JSONB, default=[])
    features_ar = Column(JSONB, default=[])
    is_active = Column(Boolean, nullable=False, default=True)
    sort_order = Column(Integer, default=0)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
